runners/samza/src/main/java/org/apache/beam/runners/samza/adapter/UnboundedSourceSystem.java

```diff
@@ -363,20 +365,32 @@ public void run() {
       private void updateWatermark() throws InterruptedException {
         final long time = System.currentTimeMillis();
         if (time - lastWatermarkTime > watermarkInterval) {
+          long watermarkMillis = Long.MAX_VALUE;
           for (UnboundedReader reader : readers) {
             final SystemStreamPartition ssp = readerToSsp.get(reader);
             final Instant currentWatermark =
                 currentWatermarks.containsKey(ssp)
                     ? currentWatermarks.get(ssp)
                     : BoundedWindow.TIMESTAMP_MIN_VALUE;
             final Instant nextWatermark = reader.getWatermark();
+            if (nextWatermark != null) {
+              watermarkMillis = Math.min(nextWatermark.getMillis(), watermarkMillis);
+            }
             if (currentWatermark.isBefore(nextWatermark)) {
               currentWatermarks.put(ssp, nextWatermark);
-              enqueueWatermark(reader);
+              if (BoundedWindow.TIMESTAMP_MAX_VALUE.isAfter(nextWatermark)) {
+                enqueueWatermark(reader);
+              } else {
+                // Max watermark has been reached for this reader.
+                enqueueMaxWatermarkAndEndOfStream(reader);
+              }
             }
           }
 
           lastWatermarkTime = time;
+          if (watermarkMillis == BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis()) {
```

Comment:
I think using previous check of nextWatermark inside the loop should be good enough.